### ISSUE TITLE
IntModule_ZF rewrite

### ISSUE THREAD

**Issue by SKolodynski** [SKolodynski/IsarMathLib#38]
In view of recent addition of integer power in groups in [IntGroup_ZF](https://github.com/SKolodynski/IsarMathLib/blob/master/IsarMathLib/IntGroup_ZF.thy) I think `IntModule_ZF` can be greatly simplified. 
@dan323, please let me know if you would like to do that, if not I will do it.

**Comment by dan323** [comment]
I can work on that, but after I go back home from the Christmas holidays. 😄

**Comment by dan323** [comment]
I saw you added a new locale as "notation". Now I do not have access to those theorems in the abelian_group locale. Locales are thought to have assumptions and definitions, and that is why they do not work well like this.

**Comment by dan323** [comment]
Since you are not defining `powz` (you are `fixing` it); when I sublocale, I loose all notation. `Fixes` implies that it should be provided from outside when the locale is used.

**Comment by SKolodynski** [comment]
[`pow`](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/Group_ZF.thy#L93) and [`nat_mult`](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/Monoid_ZF_1.thy#L58) are notations for the underlying notion of folding of a constant list, one used in the context of multiplicative notation and the other in the context of additive notation. They look similar in the source but are rendered differently at isarmathlib.org, see for example [nat_mult_add_one](https://isarmathlib.org/Monoid_ZF_1.html#a_nat_mult_add_one) and [nat_pow_add_one](https://isarmathlib.org/Group_ZF.html#a_nat_pow_add_one). The situation with `powz` is similar, it's a notation, although the additive version of this notation is not used anywhere yet. 
The idea is to rewrite `IntModule_ZF` in a locale that adds commutativity assumption to the [`group_int0`](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/IntGroup_ZF.thy#L163) locale rather than `group0` locale. I have added (temporarily) such [locale](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/IntGroup_ZF.thy#L346) and a bunch of lemmas ending with [`powz_maps_int_End`](https://isarmathlib.org/IntGroup_ZF.html#a_powz_maps_int_End) that can replace [`group_action_int_fun`](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/IntModule_ZF.thy#L154C26-L154C46) to illustrate what I have in mind.

**Comment by dan323** [comment]
You are not understanding. There is a `notation` keyword and a `definition` keyword that allows you to do those in locales. The issue is that you are doing `fixes`. When fixing, you need to provide externally, when you define or mark as a notation it forwards that when the locale is interpreted or sublocaled. For example, I could do `sublocale_name.powz` if it was defined instead of fixed. As it is now I have to rewrite the whole thing.

**Comment by SKolodynski** [comment]
Yes, I am not understanding. Practically all notation in IsarMathLib is defined in locale specifications with `fixes ... defines ... ` keywords. It does not create problems and allows me to remap all notation from one locale to a different one with `sublocale` command (which btw I learned from you). Anyway, since you say that it will be easier for you if I define `powz` with `definition` outside of locale specification I have [done](https://github.com/SKolodynski/IsarMathLib/blob/75495f50e9cccf2f09fc363b8cf1bd7ac87d54d6/IsarMathLib/IntGroup_ZF.thy#L196) that now.

**Comment by dan323** [comment]
It is different when you do topology for example, your closure operation is a definition.

**Comment by dan323** [comment]
There are things you want to remap to other things, I understand; but not pow, this has the same definition everywhere.

**Comment by SKolodynski** [comment]
As I mentioned above "[pow](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/Group_ZF.thy#L93) and [nat_mult](https://github.com/SKolodynski/IsarMathLib/blob/1e3bdc1eb00cebed7c5b35a5f60a2f15e9cbdccb/IsarMathLib/Monoid_ZF_1.thy#L58) are notations for the underlying notion of folding of a constant list". I want to be a able to prove something about $x^n = x\cdot x\cdot ... \cdot x$ in a locale where multiplicative notation is used and then map it to  a fact about $n\cdot x = x+x+ ... +x$ in a locale where additive notation is used, or vice versa - see [nat_mult_add_one](https://isarmathlib.org/Monoid_ZF_1.html#a_nat_mult_add_one) and [nat_pow_add_one](https://isarmathlib.org/Group_ZF.html#a_nat_pow_add_one) mentioned above.

**Comment by dan323** [comment]
Exactly. `pow` is defined as the fold of a constant list. That is independent of the notation of the operation used for the fold.

**Comment by SKolodynski** [comment]
So you suggest to have a notion of `pow`, defined as fold of a constant list, then in locales define various notations for it, like $n\cdot x$ named `nat_mult`  or $x^n$ named ... what?
In my view we might have a notion of fold of a constant list named say `ConstantFold` and then in locales have notations for it like  $n\cdot x$ named `nat_mult`  or $x^n$ named `pow` . I wouldn't like to name the general notion of fold of a constant list anything that associates it with "power" as people usually think of "power" as something that looks like $x^n$ and not say $n\cdot x$.

**Comment by dan323** [comment]
Please, review if this issue can be closed or something else is needed.

**Comment by dan323** [comment]
I am in Spain, so I was doing it while on a blackout 😄. I was unable to work because I could not travel to the office, so I did this. Thankfully my town got back online quite fast.

**Comment by SKolodynski** [comment]
Yes, I heard about the blackout. How did you work (I mean on IsarMathLib) without electricity? Do you have a generator?

**Comment by dan323** [comment]
The office is in Barcelona, which had no electricity and I could not use the VPN to work. My house is in a village near France, so we got it back up really quickly.

**Comment by SKolodynski** [comment]
Yes, it is sufficient. I will edit IntModule_ZF further later at some point as I would like to include it in isarmathlib.org, but that I am taking on myself. Closing the issue.
